src/Processors/Formats/ISchemaReader.cpp

```diff
@@ -120,9 +130,19 @@ NamesAndTypesList IRowWithNamesSchemaReader::readSchema()
                 type = new_type;
             /// If the new type and the previous type for this column are different,
             /// we will use default type if we have it or throw an exception.
-            else if (new_type && type->getName() != new_type->getName())
+            else if (new_type && type->equals(*new_type))
             {
-                if (default_type)
+                /// Check if we have Bool and Number and if allow_bools_as_numbers
+                /// is true make the result type Number
+                auto not_nullable_type = removeNullable(type);
+                auto not_nullable_new_type = removeNullable(new_type);
+                if (allow_bools_as_numbers && (isBool(not_nullable_type) || isBool(not_nullable_new_type))
+                    && (isNumber(not_nullable_type) || isNumber(not_nullable_new_type)))
+                {
+                    if (isBool(not_nullable_type))
+                        type = new_type;
+                }
+                else if (default_type)
                     type = default_type;
                 else
                     throw Exception(
```

Comment:
Maybe extract this logic into a private method because the logic is the same in both places.